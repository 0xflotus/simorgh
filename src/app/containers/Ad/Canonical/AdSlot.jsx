import React, { useEffect } from 'react';
import { string } from 'prop-types';
import { useLocation } from 'react-router-dom';

const AdSlot = ({ uniqueId }) => {
  const location = useLocation();
  useEffect(() => {
    window.dotcom.cmd.push(() => {
      window.dotcom.ads.registerSlot(uniqueId);
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [location]);

  return <div id={`dotcom-${uniqueId}`} className="dotcom-ad" />;
};

AdSlot.propTypes = {
  uniqueId: string.isRequired,
};

export default AdSlot;
