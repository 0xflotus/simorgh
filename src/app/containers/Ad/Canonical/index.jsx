/* eslint-disable react/jsx-curly-brace-presence */
import React, { useEffect, useState } from 'react';
import { Helmet } from 'react-helmet';
import { useLocation } from 'react-router-dom';
import styled from 'styled-components';
import {
  GEL_GROUP_3_SCREEN_WIDTH_MIN,
  GEL_GROUP_4_SCREEN_WIDTH_MIN,
} from '@bbc/gel-foundations/breakpoints';
import {
  GEL_SPACING_TRPL,
  GEL_SPACING_QUAD,
} from '@bbc/gel-foundations/spacings';
import useToggle from '#hooks/useToggle';

const StyledAd = styled.div`
  /* To centre page layout for Group 4+ */
  margin: 0 auto ${GEL_SPACING_TRPL};
  width: 100%; /* Needed for IE11 */
  text-align: center;

  @media (min-width: ${GEL_GROUP_3_SCREEN_WIDTH_MIN}) {
    margin-top: ${GEL_SPACING_TRPL};
  }

  @media (min-width: ${GEL_GROUP_4_SCREEN_WIDTH_MIN}) {
    margin-top: ${GEL_SPACING_QUAD};
    max-width: ${GEL_GROUP_4_SCREEN_WIDTH_MIN};
  }
`;

const AdSlot = ({ type }) => {
  const location = useLocation();
  useEffect(() => {
    window.dotcom.cmd.push(() => {
      window.dotcom.ads.registerSlot(type);
    });
  }, [location, type]);

  return <div id={`dotcom-${type}`} className="dotcom-ad" />;
};

const CanonicalAd = () => {
  useEffect(() => {
    window.dotcom.bootstrap({
      pageAds: true,
      playerAds: false,
    });
  }, []);

  // const { enabled: adsEnabled } = useToggle('ads');

  // if (!adsEnabled) {
  //   // return null;
  // }

  return (
    <>
      <AdSlot type="leaderboard" />
    </>
  );
};

export default CanonicalAd;
