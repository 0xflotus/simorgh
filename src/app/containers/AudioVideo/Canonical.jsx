import React from 'react';
import AudioVideo from '../../components/AudioVideo';
import deepGet from '../../lib/utilities/deepGet';
import { audioVideoPropTypes } from '../../models/propTypes';
import filterForBlockType from '../../lib/utilities/blockHandlers';
import mediaPlayerSettings from './mediaPlayerSettings';
import { RequestContext } from '../../contexts/RequestContext';

const Canonical = ({ blocks }) => {
  const { env, statsDestination, statsPageIdentifier } = React.useContext(
    RequestContext,
  );

  if (!blocks) {
    return null;
  }

  const aresMediaBlock = filterForBlockType(blocks, 'aresMedia');

  if (!aresMediaBlock) {
    return null;
  }
  const pid = deepGet(['model', 'blocks', 0, 'model', 'id'], aresMediaBlock);
  const id = `mp#${pid}`;

  return (
    <AudioVideo
      idArray={['mediaPlayer1', 'mediaPlayer2', 'mediaPlayer3']}
      id={id}
      mediaPlayerSettings={mediaPlayerSettings({
        aresMediaBlock,
        env,
        pid,
        statsDestination,
        statsPageIdentifier,
      })}
      width="100%"
      height="26em"
    />
  );
};

Canonical.propTypes = {
  ...audioVideoPropTypes,
};

export default Canonical;
