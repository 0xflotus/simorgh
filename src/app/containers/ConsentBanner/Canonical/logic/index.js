import 'isomorphic-fetch';
import Cookie from 'js-cookie';
import cookieOvenUrl from './cookieOvenUrl';

const PRIVACY_COOKIE = 'ckns_privacy';
const EXPLICIT_COOKIE = 'ckns_explicit';
const POLICY_COOKIE = 'ckns_policy';
const COOKIE_EXPIRY = 365;
const BANNER_APPROVED = '1';
const POLICY_APPROVED = '111';
const POLICY_DENIED = '000';

const onClient = typeof window !== 'undefined';

const setCookie = (name, value) =>
  Cookie.set(name, value, { expires: COOKIE_EXPIRY });

const setCookieOven = (value, logger) => {
  if (window.location && window.location.origin) {
    try {
      fetch(
        `${cookieOvenUrl(window.location.origin)}/${POLICY_COOKIE}/${value}`,
      );
    } catch (e) {
      const log = logger || console;
      log.error(e);
    }
  }
};

const setPolicyCookie = (value, logger) => {
  setCookie(POLICY_COOKIE, value);
  setCookieOven(value, logger);
};

const seenPrivacyBanner = () => Cookie.get(PRIVACY_COOKIE) === BANNER_APPROVED;
const seenCookieBanner = () => Cookie.get(EXPLICIT_COOKIE) === BANNER_APPROVED;
const policyCookieSet = () => !!Cookie.get(POLICY_COOKIE);

const setSeenPrivacyBanner = () => setCookie(PRIVACY_COOKIE, BANNER_APPROVED);
const setDefaultPolicy = logger => setPolicyCookie(POLICY_DENIED, logger);
const setAppovedPolicy = logger => setPolicyCookie(POLICY_APPROVED, logger);
const setDismissedCookieBanner = () =>
  setCookie(EXPLICIT_COOKIE, BANNER_APPROVED);

const canonicalBannerLogic = ({ showPrivacyBanner, showCookieBanner, logger }) => {
  const runInitial = () => {
    if (onClient) {
      if (!seenPrivacyBanner()) {
        showPrivacyBanner(true);
        setSeenPrivacyBanner();
      }

      if (!seenCookieBanner()) {
        showCookieBanner(true);
      }

      if (!policyCookieSet()) {
        setDefaultPolicy(logger);
      }
    }
  };

  const privacyOnAllow = () => {
    showPrivacyBanner(false);
  };

  const privacyOnReject = () => {
    showPrivacyBanner(false);
  };

  const cookieOnAllow = () => {
    showCookieBanner(false);
    setDismissedCookieBanner();
    setAppovedPolicy(logger);
  };

  const cookieOnReject = () => {
    showCookieBanner(false);
    setDismissedCookieBanner();
  };

  return {
    runInitial,
    privacyOnAllow,
    privacyOnReject,
    cookieOnAllow,
    cookieOnReject,
  };
};

export default canonicalBannerLogic;
