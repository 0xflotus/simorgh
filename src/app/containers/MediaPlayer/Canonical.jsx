import React, { useState } from 'react';
import styled from 'styled-components';
import Image from '@bbc/psammead-image';

const IMG_SRC =
  'https://ichef.bbci.co.uk/news/640/cpsprodpb/E7DB/production/_101655395_paulineclayton.jpg';

const StyledPlaceholder = styled.div`
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
`;

const StyledIframe = styled.iframe`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border: 0;
  overflow: hidden;
`;

const MediaPlayer = () => (
  <StyledIframe
    title="Media Player"
    src="https://www.test.bbc.co.uk/ws/av-embeds/c3wmq4d1y3wo/p01k6msp?morph_env=test"
    scrolling="no"
    allowFullScreen
  />
);

const Canonical = () => {
  const [showMediaPlayer, setShowMediaPlayer] = useState(false);
  const handlePlaceholderClick = () => setShowMediaPlayer(true);

  return showMediaPlayer ? (
    <MediaPlayer />
  ) : (
    <StyledPlaceholder onClick={handlePlaceholderClick}>
      <Image alt="Image Alt" src={IMG_SRC} />
    </StyledPlaceholder>
  );
};
export default Canonical;
