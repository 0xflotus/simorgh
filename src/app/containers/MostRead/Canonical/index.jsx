import React, { useEffect, useState, useContext } from 'react';

import 'isomorphic-fetch';
import { string } from 'prop-types';
import {
  MostReadList,
  MostReadItemWrapper,
  MostReadRank,
  MostReadLink,
} from '@bbc/psammead-most-read';
import SectionLabel from '@bbc/psammead-section-label';
import { ServiceContext } from '#contexts/ServiceContext';
import webLogger from '#lib/logger.web';
import { mostReadRecordIsFresh, shouldRenderLastUpdated } from '../utilities';
import LastUpdated from './LastUpdated';

const logger = webLogger();

const CanonicalMostRead = ({ endpoint }) => {
  const [items, setItems] = useState([]);
  const {
    service,
    script,
    dir,
    datetimeLocale,
    timezone,
    mostRead: { header, lastUpdated, numberOfItems },
  } = useContext(ServiceContext);

  useEffect(() => {
    const handleResponse = async response => {
      const mostReadData = await response.json();
      const isLocalEnv = process.env.SIMORGH_APP_ENV === 'local';
      // do not show most read if lastRecordUpdated is greater than 35min as this means PopAPI has failed twice
      // in succession. This suggests ATI may be having issues, hence risk of stale data.
      if (
        isLocalEnv ||
        mostReadRecordIsFresh(mostReadData.lastRecordTimeStamp)
      ) {
        const mostReadItems = mostReadData.records
          .slice(0, numberOfItems)
          .map(({ id, promo: { headlines, locators, timestamp } }) => ({
            id,
            title: headlines.shortHeadline,
            href: locators.assetUri,
            timestamp: shouldRenderLastUpdated(timestamp) && (
              <LastUpdated
                prefix={lastUpdated}
                script={script}
                service={service}
                timestamp={timestamp}
                locale={datetimeLocale}
                timezone={timezone}
              />
            ),
          }));
        setItems(mostReadItems);
      }
    };
    const fetchMostReadData = pathname =>
      fetch(pathname, { mode: 'no-cors' })
        .then(handleResponse)
        .catch(e => logger.error(`HTTP Error: "${e}"`));
    fetchMostReadData(endpoint);
  }, [endpoint, numberOfItems, datetimeLocale, lastUpdated, script, service]);

  return (
    <>
      {items.length ? (
        // eslint-disable-next-line jsx-a11y/no-redundant-roles
        <section role="region" aria-labelledby="Most-Read">
          <SectionLabel
            script={script}
            labelId="Most-Read"
            service={service}
            dir={dir}
          >
            {header}
          </SectionLabel>
          <MostReadList numberOfItems={items.length} dir={dir}>
            {items.map((item, i) => (
              <MostReadItemWrapper dir={dir} key={item.id}>
                <MostReadRank
                  service={service}
                  script={script}
                  listIndex={i + 1}
                  numberOfItems={items.length}
                  dir={dir}
                />
                <MostReadLink
                  dir={dir}
                  service={service}
                  script={script}
                  title={item.title}
                  href={item.href}
                >
                  {item.timestamp}
                </MostReadLink>
              </MostReadItemWrapper>
            ))}
          </MostReadList>
        </section>
      ) : null}
    </>
  );
};

CanonicalMostRead.propTypes = {
  endpoint: string.isRequired,
};

export default CanonicalMostRead;
