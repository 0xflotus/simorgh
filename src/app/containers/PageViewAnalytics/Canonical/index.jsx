import { useEffect, useContext } from 'react';
import { shape } from 'prop-types';
import articlePropTypes from '../../../models/propTypes/article';
import { ServiceContext } from '../../../contexts/ServiceContext';
import { RequestContext } from '../../../contexts/RequestContext';
import getPageViewBeaconUrl from './getPageViewBeaconUrl';
import deepGet from '../../../helpers/json/deepGet';
import onClient from '../../../helpers/onClient';
import sendBeacon from '../../../lib/analyticsUtils/sendBeacon';

const CanonicalPageViewAnalytics = ({ articleData }) => {
  const { platform, isUK, env } = useContext(RequestContext);

  if (onClient()) {
    const { service } = useContext(ServiceContext);
    const optimoId = deepGet(['promo', 'id'], articleData);

    // Only send page beacon when optimoId changes
    useEffect(() => {
      const pageViewBeaconUrl = getPageViewBeaconUrl({
        articleData,
        service,
        platform,
        isUK,
        env,
      });

      sendBeacon(pageViewBeaconUrl);
    }, [optimoId]);
  }

  return null;
};

CanonicalPageViewAnalytics.propTypes = {
  articleData: shape(articlePropTypes).isRequired,
};

export default CanonicalPageViewAnalytics;
