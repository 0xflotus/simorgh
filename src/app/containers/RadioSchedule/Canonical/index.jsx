import React, { useEffect, useState, useContext } from 'react';
import 'isomorphic-fetch';
import { string } from 'prop-types';
import styled from 'styled-components';
import {
  GEL_GROUP_2_SCREEN_WIDTH_MAX,
  GEL_GROUP_3_SCREEN_WIDTH_MIN,
  GEL_GROUP_3_SCREEN_WIDTH_MAX,
  GEL_GROUP_4_SCREEN_WIDTH_MIN,
} from '@bbc/gel-foundations/breakpoints';
import {
  GEL_SPACING_DBL,
  GEL_SPACING_TRPL,
  GEL_SPACING_QUAD,
} from '@bbc/gel-foundations/spacings';
import RadioSchedule from '@bbc/psammead-radio-schedule';
import moment from 'moment';
import findLastIndex from 'ramda/src/findLastIndex';
import propSatisfies from 'ramda/src/propSatisfies';
import SectionLabel from '@bbc/psammead-section-label';
import { Link } from '@bbc/psammead-story-promo';
import { ServiceContext } from '#contexts/ServiceContext';
import webLogger from '#lib/logger.web';

const logger = webLogger();

const MarginWrapper = styled.div`
  @media (max-width: ${GEL_GROUP_2_SCREEN_WIDTH_MAX}) {
    margin-top: ${GEL_SPACING_DBL};
  }
  @media (min-width: ${GEL_GROUP_3_SCREEN_WIDTH_MIN}) {
    margin-top: ${GEL_SPACING_TRPL};
  }
  @media (max-width: ${GEL_GROUP_3_SCREEN_WIDTH_MAX}) {
    margin-bottom: ${GEL_SPACING_DBL};
  }
  @media (min-width: ${GEL_GROUP_4_SCREEN_WIDTH_MIN}) {
    margin-bottom: ${GEL_SPACING_TRPL};
  }
`;

const RadioScheduleSection = styled.section.attrs(() => ({
  role: 'region',
  'aria-labelledby': 'Radio-Schedule',
}))``;

const FrontPageRadioScheduleSection = styled(RadioScheduleSection)`
  margin: 0 auto;
`;

const RadioScheduleSectionLabel = styled(SectionLabel)`
  @media (min-width: ${GEL_GROUP_4_SCREEN_WIDTH_MIN}) {
    margin-top: ${GEL_SPACING_QUAD};
  }
`;

const RadioFrequencyLink = styled(Link)`
  font-size: 14px;
  line-height: 18px;
`;

const CanonicalRadioSchedule = ({ endpoint }) => {
  const [schedule, setRadioSchedule] = useState();
  const {
    service,
    script,
    dir,
    timezone,
    locale,
    radioSchedule: { header, frequenciesPageUrl, frequenciesPageLabel },
  } = useContext(ServiceContext);

  const getProgramState = (currentTime, startTime, endTime) => {
    const isLive = currentTime < endTime && currentTime > startTime;
    if (isLive) {
      return 'live';
    }
    const hasEnded = currentTime > endTime;
    if (hasEnded) {
      return 'onDemand';
    }
    return 'next';
  };

  useEffect(() => {
    const getLink = (state, program) => {
      const url = `/${service}/${program.serviceId}`;
      return state === 'live'
        ? `${url}/liveradio`
        : `${url}/${program.broadcast.pid}`;
    };

    const handleResponse = async response => {
      const radioScheduleData = await response.json();

      const currentTime = parseInt(moment.utc().format('x'), 10);
      // finding latest program, that may or may not still be live. this is because there isn't
      // always a live program, in which case we show the most recently played program on demand.
      const latestProgramIndex = findLastIndex(
        propSatisfies(time => time < currentTime, 'publishedTimeStart'),
      )(radioScheduleData.schedules);

      const radioSchedules = radioScheduleData.schedules;

      const scheduleDataIsComplete =
        radioSchedules[latestProgramIndex - 2] &&
        radioSchedules[latestProgramIndex + 1];
      const schedulesToShow = scheduleDataIsComplete && [
        radioSchedules[latestProgramIndex],
        radioSchedules[latestProgramIndex - 1],
        radioSchedules[latestProgramIndex - 2],
        radioSchedules[latestProgramIndex + 1],
      ];

      const schedules =
        schedulesToShow &&
        schedulesToShow.map(program => {
          const currentState = getProgramState(
            currentTime,
            program.publishedTimeStart,
            program.publishedTimeEnd,
          );
          return {
            id: program.broadcast.pid,
            state: currentState,
            stateLabel: currentState,
            startTime: program.publishedTimeStart,
            link: getLink(currentState, program),
            brandTitle: program.brand.title,
            episodeTitle: program.episode.presentationTitle,
            summary: program.episode.synopses.short,
            duration: program.publishedTimeDuration,
            durationLabel: 'Duration',
          };
        });
      setRadioSchedule(schedules);
    };

    const fetchRadioScheduleData = pathname =>
      fetch(pathname, { mode: 'no-cors' })
        .then(handleResponse)
        .catch(e => logger.error(`HTTP Error: "${e}"`));

    fetchRadioScheduleData(endpoint);
  }, [endpoint, service, script, timezone, locale]);

  if (!schedule) {
    return null;
  }

  const renderFrequencyLink = (link, label) => (
    <RadioFrequencyLink link={link}>{label}</RadioFrequencyLink>
  );

  return (
    <FrontPageRadioScheduleSection>
      <RadioScheduleSectionLabel
        script={script}
        labelId="Radio-Schedule"
        service={service}
        dir={dir}
        bar={false}
      >
        {header}
      </RadioScheduleSectionLabel>
      <MarginWrapper>
        <RadioSchedule
          schedules={schedule}
          locale={locale}
          timezone={timezone}
          script={script}
          service={service}
          dir={dir}
        />
        {frequenciesPageUrl &&
          renderFrequencyLink(frequenciesPageUrl, frequenciesPageLabel)}
      </MarginWrapper>
    </FrontPageRadioScheduleSection>
  );
};

CanonicalRadioSchedule.propTypes = {
  endpoint: string.isRequired,
};

export default CanonicalRadioSchedule;
