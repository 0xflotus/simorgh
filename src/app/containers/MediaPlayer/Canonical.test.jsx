import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Canonical from './Canonical';

describe('Media Player: Canonical', () => {
  it('should render a placeholder image', () => {
    const tree = renderer.create(
      <Canonical
        embedSrc="https://foo.bar/"
        placeholderSrc="https://foo.bar/img"
      />,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should render a MediaPlayer component when placeholder is clicked', () => {
    const component = mount(
      <Canonical embedSrc="http://foo.com" placeholderSrc="http://bar.com" />,
    );

    component.simulate('click');

    expect(component).toMatchSnapshot();
  });
});
