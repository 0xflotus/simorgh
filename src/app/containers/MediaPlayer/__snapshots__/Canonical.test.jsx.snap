// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Media Player: Canonical should render a MediaPlayer component when placeholder is clicked 1`] = `
<Canonical
  embedSrc="http://foo.com"
  placeholderSrc="http://bar.com"
>
  <MediaPlayer
    src="http://foo.com"
  >
    <StyledIframe
      allow="autoplay; fullscreen"
      gesture="media"
      scrolling="no"
      src="http://foo.com"
      title="Media Player"
    >
      <StyledComponent
        allow="autoplay; fullscreen"
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "StyledIframe-s82gr2-1",
              "isStatic": true,
              "lastClassName": "nVHgH",
              "rules": Array [
                "position:absolute;top:0;left:0;width:100%;height:100%;border:0;overflow:hidden;",
              ],
            },
            "displayName": "StyledIframe",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "StyledIframe-s82gr2-1",
            "target": "iframe",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        gesture="media"
        scrolling="no"
        src="http://foo.com"
        title="Media Player"
      >
        <iframe
          allow="autoplay; fullscreen"
          className="StyledIframe-s82gr2-1 nVHgH"
          scrolling="no"
          src="http://foo.com"
          title="Media Player"
        />
      </StyledComponent>
    </StyledIframe>
  </MediaPlayer>
</Canonical>
`;

exports[`Media Player: Canonical should render a placeholder image 1`] = `
<div
  className="StyledPlaceholder-s82gr2-0 cFOMxr"
  onClick={[Function]}
>
  <img
    alt="Image Alt"
    className="StyledImg-sc-7vx2mr-0 hIxkbt"
    height={null}
    sizes={null}
    src="https://foo.bar/img"
    srcSet={null}
    width={null}
  />
</div>
`;
